consensus/src/block_storage/sync_manager.rs

```diff
@@ -231,45 +233,68 @@ impl BlockStore {
             + 1;
 
         let blocks = retriever
-            .retrieve_block_for_qc(highest_ordered_cert, num_blocks)
+            .retrieve_block_for_qc(
+                highest_ordered_cert,
+                num_blocks,
+                highest_ledger_info.commit_info().id(),
+            )
             .await?;
 
         assert_eq!(
             blocks.first().expect("should have at least 3-chain").id(),
             highest_ordered_cert.certified_block().id(),
+            "Expecting in the retrieval response, first block should be {}, but got {}",
+            highest_ordered_cert.certified_block().id(),
+            blocks.first().expect("should have at least 3-chain").id(),
         );
 
         assert_eq!(
-            blocks.last().expect("should have at least 3-chain").id(),
-            highest_ledger_info.commit_info().id(),
-        );
-
-        // although unlikely, we might wrap num_blocks around on a 32-bit machine
-        assert!(num_blocks < std::usize::MAX as u64);
-        let mut quorum_certs = vec![highest_ordered_cert.clone()];
-        quorum_certs.extend(
             blocks
-                .iter()
-                .take(num_blocks as usize - 1)
-                .map(|block| block.quorum_cert().clone()),
+                .first()
+                .expect("should have at least 3-chain")
+                .round(),
+            highest_ordered_cert.certified_block().round(),
+            "Expecting in the retrieval response, first block round should be {}, but got round {}",
+            highest_ordered_cert.certified_block().round(),
+            blocks
+                .first()
+                .expect("should have at least 3-chain")
+                .round(),
         );
-        for (i, block) in blocks.iter().enumerate() {
-            assert_eq!(block.id(), quorum_certs[i].certified_block().id());
-        }
 
-        // If a node restarts in the middle of state synchronization, it is going to try to catch up
-        // to the stored quorum certs as the new root.
-        storage.save_tree(blocks.clone(), quorum_certs.clone())?;
-        state_computer.sync_to(highest_ledger_info).await?;
+        if blocks
+            .iter()
+            .any(|b| b.id() == highest_ledger_info.ledger_info().commit_info().id())
```

Comment:
we should be able to directly use the first block if the request (with target id) succeeds.